### ISSUE TITLE
make inpaint opposed the default algorithm

### ISSUE THREAD

**Issue by TurboGit** [darktable-org/darktable#12650]
This is a follow up for #12639.

The regression has been solved by setting back the default highlights recovery algorithm to CLIP.

We want the new INPANT OPPOSED to be the default because it rocks.

Fact is that is not easy at all here. We have some old XMP not having the highlights module listed as it was using the default parameters. After lot of thoughts about this I think the only solution is to go the hard way.

Proposed solution:
1. in develop.c : _dev_auto_apply_presets()
2. check that hightlights is already applied
3. if not add hightlights with default params (calling _dev_insert_module)

I didn't test this, my fear is that it could already be too late... But this is worth a try.

@jenshannoschwalm : Let me know if you want to try this otherwise I can check.

In any case, we'll discuss this issue together I really don't want to miss the inpaint as default for 4.2.

If any reader have a idea about this, please help !

**Comment by jenshannoschwalm** [comment]
If we could check the xmp version at runtime  knowing "bad ones" without hl module included it would be probably easy to set mode by force in hl module.

**Comment by TurboGit** [comment]
We can't do that @jenshannoschwalm, there is no version being bad or some being good. The bad ones are all those that have an old version without having changed the default highlights recovery mode. That's why using the some xmp version won't help.

**Comment by TurboGit** [comment]
In the regression tests the first 90 tests are all having the xmp_version = 4 but only ~20 are failing (no highlights recovery recorded in the xmp).

**Comment by parafin** [comment]
Is there any other iops not being recorded in the history? I thought there was a consensus that everything should go there, there should be a ticket about it. Anyway, I guess we need a flag in xmp like legacy_presets_applied (but named like legacy_hl_recovery_applied), which would indicate whether old clip hl recovery should be added to the history.

**Comment by jenshannoschwalm** [comment]
I think sharpen has the same issue.

**Comment by TurboGit** [comment]
> I thought there was a consensus that everything should go there,

Yes, and that's the case now but the issue is for old XMP.

**Comment by parafin** [comment]
Well, can’t default be changed only for new history stacks? Like it’s done with preferences setting about modern/legacy WB etc.

**Comment by jenshannoschwalm** [comment]
Good idea, i certainly wouldn't want more options but what about binding the "opposed" as default to modern?

**Comment by gi-man** [comment]
I like the idea of biding to modern. 

I would also think we should change to make modern the default instead of legacy.

**Comment by TurboGit** [comment]
Note that we have the same issue on library.db where old edits do not have highlights in the history, so we also need a way to ensure that such edits do not get the new inpaint default.

**Comment by TurboGit** [comment]
> Well, can’t default be changed only for new history stacks? 

Yes, using `reload_default()` but this seems too late when loading XMP. I have tried this approach without good result.

**Comment by gi-man** [comment]
What about using the Has edits tag? If no edits, then default to inpatient, otherwise don't touch it?

**Comment by TurboGit** [comment]
> What about using the Has edits tag? If no edits, then default to inpatient, otherwise don't touch it?

Ah yes good idea, could works indeed, worth a try anyway.

**Comment by parafin** [comment]
> > Well, can’t default be changed only for new history stacks?
> 
> Yes, using `reload_default()` but this seems too late when loading XMP. I have tried this approach without good result.

I don’t understand how the situation differs from other default setting switches we have in preferences…

**Comment by jenshannoschwalm** [comment]
I tested this so far in highlights.c.

```
void reload_defaults(dt_iop_module_t *module)
{
  // we might be called from presets update infrastructure => there is no image
  if(!module->dev || module->dev->image_storage.id == -1) return;

  dt_iop_highlights_params_t *d = module->default_params;
  const gboolean is_modern = dt_conf_is_equal("plugins/darkroom/chromatic-adaptation", "modern");
  if(is_modern)
    d->mode = DT_IOP_HIGHLIGHTS_OPPOSED;
  else
    d->mode = DT_IOP_HIGHLIGHTS_CLIP;
.....
```
What do you think?

**Comment by TurboGit** [comment]
This won't work if you open an image having an XMP without highlights (old edit) and wither darktable set to "modern".

**Comment by jenshannoschwalm** [comment]
Yes, also requiring checks as you suggested in develop.c

But would you agree on binding to modern?

**Comment by jenshannoschwalm** [comment]
I mean I would do this as for temperature?

**Comment by TurboGit** [comment]
I'm not fan to bind this to modern as this is not related. The legacy vs modern was for the white balance (temperature ws color calibration). I would find sad to not have inpaint by default in legacy.

**Comment by jenshannoschwalm** [comment]
Ok. We certainly need some sort of switch flag in conf like "bool modern highlights" with default to true resulting  in opposed checked in highlights default parameters. 

In develop.c we check somewhat like for temperature and insert with above flag set to false thus clip.

Would that work?

**Comment by TurboGit** [comment]
Why not update the highlight module version and makes inpaint the default for the new version only? In legacy params we always fallback on CLIP default if not set otherwise. This will solve old edits issue. Only XMP still have to be fixed another way.

**Comment by TurboGit** [comment]
> In develop.c we check somewhat like for temperature and insert with above flag set to false thus clip.

This may be the solution for old edits in XMP.

**Comment by TurboGit** [comment]
> Why not update the highlight module version and makes inpaint the default for the new version only?

We also need a new boolean in params set to TRUE for old edits, that is the ones that have gone through `legacy_params()`.

**Comment by jenshannoschwalm** [comment]
I will have a look. If you know already how to do it, go ahead 😀
